src/Helpers/DefinitionValidator.php

```diff
@@ -76,108 +101,243 @@ public static function validateArrayDefinition(array $definition, ?string $id =
-                    throw new InvalidConfigException(
-                        sprintf(
-                            'Invalid definition: incorrect method name. Got "%s".',
-                            $key
-                        )
-                    );
-                }
-                if (!is_array($value)) {
-                    throw new InvalidConfigException(
-                        sprintf(
-                            'Invalid definition: incorrect method arguments. Expected array, got %s.',
-                            get_debug_type($value)
-                        )
-                    );
-                }
+                self::validateMethod($key, $classReflection, $classPublicMethods, $className, $value);
                 continue;
             }
             if (str_starts_with($key, '$')) {
+                self::validateProperty($key, $classReflection, $classPublicProperties, $className);
                 continue;
             }
 
-            self::throwInvalidArrayDefinitionKey($key);
-        }
+            $possibleOptionsMessage = self::generatePossibleMessage(
+                $key,
+                $classPublicMethods,
+                $classPublicProperties,
+                $classReflection,
+                $className
+            );
 
-        if ($id === null && !isset($definition[ArrayDefinition::CLASS_NAME])) {
-            throw new InvalidConfigException('Invalid definition: no class name specified.');
+            throw new InvalidConfigException(
+                "Invalid definition: key \"$key\" is not allowed. $possibleOptionsMessage",
+            );
         }
     }
 
     /**
-     * @throws InvalidConfigException
+     * Deny `DefinitionInterface`, exclude `ReferenceInterface`
      */
-    private static function throwInvalidArrayDefinitionKey(string $key): void
+    private static function isValidObject(object $value): bool
     {
-        $preparedKey = trim(strtr($key, [
-            '()' => '',
-            '$' => '',
-        ]));
+        return !($value instanceof DefinitionInterface) || $value instanceof ReferenceInterface;
+    }
 
-        if ($preparedKey === '' || !preg_match('/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*$/', $preparedKey)) {
+    private static function validateClassName(mixed $class): void
+    {
+        if (!is_string($class)) {
             throw new InvalidConfigException(
-                sprintf('Invalid definition: key "%s" is not allowed.', $key)
+                sprintf(
+                    'Invalid definition: class name must be a non-empty string, got %s.',
+                    get_debug_type($class),
+                )
             );
         }
+        if (trim($class) === '') {
+            throw new InvalidConfigException(
+                sprintf(
+                    'Invalid definition: class name must be a non-empty string.',
+                )
+            );
+        }
+        if (!class_exists($class)) {
```

Comment:
?

Reply:
We can use something old library that don't use namespaces and run `require ...` after container initializiation.

Reply:
Waiting for an example for your case

Reply:
[JpGraph](https://jpgraph.net/download/) don't use namespaces and can be connected (via `required ...`) after container configuration.

Reply:
`class_exists` triggers autoload. That could affect overall performance significantly.

Reply:
> `class_exists` triggers autoload. That could affect overall performance significantly.

In production environment validation should be disabled. So performance not actual here.

Reply:
Then it's fine. Loading on-demand via non-autoload is rare enough to ignore it nowadays.

Reply:
I don't like the way without `class_exists` in case when all of your definitions must be validated.
If validation fails and without validation it works it's not validation's problem.

Reply:
If validation fails and without validation it works it's incorrect validation.

Reply:
Also, performance effect could be very significant even for debug mode.